Added texture support

Mercesa opened the pull request:
Added texture support
Resource management ( still in only functional phase )

Mercesa committed:
Added support for one texture loading

Still very hacky, will improve in future commit

Mercesa committed:
Fixed all the issues

Fixed memory leaks by using unique ptrs, a bit unwise to use smart
pointers if I opt for a single ownership system.

Also added comptrs, was an issue when getting a reference of a pointer.
By creating a contemporary pointer it takes the adress of this new
pointer instead of the original one. Uses get address of comptr

Evayr commented on DirectX11Proj/Application.cpp:
```diff
@@ -14,8 +14,8 @@ Application::~Application()
 { 
-	auto tScene = new PlayerSceneExample();
-	LoadScene(tScene);
+	auto tScene = std::make_unique<PlayerSceneExample>();
```
t as in temp? tpScene

Evayr replied:
p because still a pointer

Mercesa replied:
resolved

Evayr commented on DirectX11Proj/IApplication.cpp:
```diff
@@ -2,47 +2,46 @@
 
-IApplication::IApplication() : mpCurrentScene(nullptr), mShouldQuit(false)
+
+IApplication::IApplication() : mShouldQuit(false)
```
may want to do the : mShouldQuit(false) on a new line. Up to you.

Mercesa replied:
resolved

Evayr commented on DirectX11Proj/IApplication.cpp:
```diff
@@ -2,47 +2,46 @@
+void IApplication::LoadScene(std::unique_ptr<IScene> apScene)
 {
-	assert(aScene != nullptr);
+	assert(apScene != nullptr);
```
or just assert(apScene)

Evayr commented on DirectX11Proj/IApplication.cpp:
```diff
@@ -2,47 +2,46 @@
 	{
-		std::cout << "There is already a scene loaded!" << std::endl;
-		return;
+		mpCurrentScene.release();
```
if you do mpCurrentScene = std::move(apScene) then the old value is automatically released.

Mercesa replied:
resolved

Evayr commented on DirectX11Proj/IApplication.cpp:
```diff
@@ -2,47 +2,46 @@
 	}
 
-	this->mpCurrentScene = aScene;
+	mpCurrentScene = std::move(apScene);;
```
two semicolons;;

Mercesa replied:
resolved

Evayr commented on DirectX11Proj/IApplication.cpp:
```diff
@@ -2,47 +2,46 @@
-		aScene->Init();
-		aScene->mInitialized = true;
+		mpCurrentScene->Init();
+		mpCurrentScene->mInitialized = true;
```
The Init() function should set mInitialized to true. The application shouldn't do that for the Scene.

Mercesa replied:
It's a bit hard because I do not want to put this burden on the user, unless it calls the virtual function as well when I call the overidden function? I do not know

Evayr replied:
Maybe don't make the ctor virtual, and only initialize the core things such as mInitialized. The rest can be done by the user

Evayr commented on DirectX11Proj/IScene.cpp:
```diff
@@ -3,9 +3,9 @@
 
 // Create camera on creation of scene
-IScene::IScene()
+IScene::IScene() : mInitialized(false)
```
if you do decide to go with the fast declaration on a new line, change this one as well

Mercesa replied:
fixed

Evayr commented on DirectX11Proj/IScene.cpp:
```diff
@@ -3,9 +3,9 @@
+IScene::IScene() : mInitialized(false)
 {
-	mCamera = std::make_shared<CameraClass>();
+	mCamera = std::make_unique<CameraClass>();
```
mpCamera

Mercesa replied:
resolved

Evayr commented on DirectX11Proj/IScene.h:
```diff
@@ -17,19 +17,19 @@ class IScene
 
 	// Just a bit lazy and turn it into a vector, will convert back to a list later
-	std::vector <std::shared_ptr<IObject>> mObjects;
+	std::vector <std::unique_ptr<IObject>> mObjects;
```
std::list

Mercesa replied:
http://stackoverflow.com/questions/238008/relative-performance-of-stdvector-vs-stdlist-vs-stdslist

Lists has its uses but this question and the people answering elaborates a bit more on my decisions

Evayr commented on DirectX11Proj/IScene.h:
```diff
@@ -17,19 +17,19 @@ class IScene
 
 protected:
-	std::shared_ptr<CameraClass> mCamera;
+	std::unique_ptr<CameraClass> mCamera;
```
mpCamera

Mercesa replied:
fixed

Evayr commented on DirectX11Proj/IScene.h:
```diff
@@ -17,19 +17,19 @@ class IScene
+	std::unique_ptr<CameraClass> mCamera;
 
 private:
 	bool mInitialized = false;
```
initialize this in the constructor

Mercesa replied:
resolved

Evayr commented on DirectX11Proj/ModelData.h:
```diff
@@ -26,6 +26,6 @@ struct MeshData
 	std::vector<uint32_t> indices;
 
-	std::string textureFilePath;
+	std::string textureFilePath = "empty";
```
initialize this in the ctor

Evayr replied:
Also maybe use "" instead of "empty". What is a texture is called "empty"?

Mercesa replied:
Since it's a struct I'm leaving it at this initialization, but the reason I chose empty is because it is easy to identify if a texture is empty or not. I will consider "" and have to reconsider it from a debugging perspective.

Evayr commented on DirectX11Proj/ModelLoader.cpp:
```diff
@@ -105,6 +105,7 @@ std::string GetTextureLocation(aiMaterial* const a_Mat, aiTextureType a_Type)
 	// Check the amount of textures of a specific type
 	if ((count = a_Mat->GetTextureCount(a_Type)) <= 0)
 	{
+		std::cout << "no texture detected fam" << std::endl;
```
n1 comment m8

Mercesa replied:
resolved

Evayr commented on DirectX11Proj/PlayerSceneExample.cpp:
```diff
@@ -14,29 +14,29 @@ PlayerSceneExample::~PlayerSceneExample()
 	int mouseRelX = 0;
 	int mouseRelY = 0;
-	aInput->GetMouseRelLocation(mouseRelX, mouseRelY);
+	apInput->GetMouseRelLocation(mouseRelX, mouseRelY);
```
what is "Rel"Location? Relative? Relocation? Release? I'd write it out in full

Mercesa replied:
fair enough changed and resolved

Evayr commented on DirectX11Proj/PlayerSceneExample.cpp:
```diff
@@ -14,29 +14,29 @@ PlayerSceneExample::~PlayerSceneExample()
 	mCamera->SetRotation(y, x, 0);
 
-	if (aInput->mKeyboardState[DIK_W] & 0x80)
+	if (apInput->mKeyboardState[DIK_W] & 0x80)
```
replace 0x80 with an enum in the .h file. Like this:

public:
   enum
   {
      kWFlag = 0x80, // or something
   }

   Class();
   ~Class();

Mercesa replied:
will put it on my todo

Evayr commented on DirectX11Proj/PlayerSceneExample.cpp:
```diff
@@ -52,10 +52,9 @@ void PlayerSceneExample::Init()
 	for (int i = 0; i < tModels.size(); ++i)
 	{
-		std::shared_ptr<ObjectExample> tObject = std::make_shared<ObjectExample>();
+		std::unique_ptr<ObjectExample> tObject = std::make_unique<ObjectExample>();
```
tpObject

Mercesa replied:
fixed

Evayr commented on DirectX11Proj/PlayerSceneExample.cpp:
```diff
@@ -52,10 +52,9 @@ void PlayerSceneExample::Init()
 		tObject->mpModel = tModels[i];
-		this->mObjects.push_back(tObject);
-		std::cout << "Going through models" << std::endl;
+		this->mObjects.push_back(std::move(tObject));
```
remove this->

Evayr replied:
maybe add a separate function to add objects to a scene, like AddObject()

Mercesa replied:
Will put it on my todo, not sure how to go about it properly

Evayr commented on DirectX11Proj/ResourceManager.cpp:
```diff
@@ -1,29 +1,65 @@
+void ResourceManager::Shutdown()
+{
+	this->mLoadedModels.clear();
+	this->mLoadedTextures.clear();
```
remove this->

Also you may want to create a dtor that calls Shutdown()

Mercesa replied:
fair enough, will think about it

Evayr commented on DirectX11Proj/ResourceManager.cpp:
```diff
@@ -1,29 +1,65 @@
-	
-	for (unsigned int i = 0; i < mpModelLoader->GetMeshesToBeProcessed().size(); ++i)
+	std::unique_ptr<ModelClass> tModelClass = nullptr;
+	std::unique_ptr<d3dTexture> tTextureClass = nullptr;
```
tpTextureClass, tpModelClass

Mercesa replied:
fixed

Evayr commented on DirectX11Proj/ResourceManager.cpp:
```diff
@@ -1,29 +1,65 @@
+		{
+			LOG(WARNING) << "ModelLoading: Finished loading model " << aFilePath;
+		}
+
```
may want to remove the enter between the if and else, so you know they're both related to each other and no accidental code can get in between

Mercesa replied:
fixed

Evayr commented on DirectX11Proj/ResourceManager.cpp:
```diff
@@ -1,29 +1,65 @@
+		}
+
 		tModels.push_back(tModelClass.get());
 		mLoadedModels.push_back(std::move(tModelClass));
```
Good use. I like seeing this.

Evayr commented on DirectX11Proj/ResourceManager.h:
```diff
@@ -26,7 +28,6 @@ class ResourceManager
 	}
 
-
 	std::vector<ModelClass*> LoadModels(std::string aFilePath);
```
If people are loading in models dynamically throughout the gameplay, you may want to consider using std::list instead

Mercesa replied:
fair, since we don't access models as frequently

Evayr commented on DirectX11Proj/ResourceManager.h:
```diff
@@ -35,7 +36,10 @@ class ResourceManager
-	std::vector<std::shared_ptr<ModelClass>> mLoadedModels;
+	// Turn these into maps at one point
+	std::vector<std::unique_ptr<ModelClass>> mLoadedModels;
+	std::vector<std::unique_ptr<d3dTexture>> mLoadedTextures;
```
Just want to say that you did the right thing by not naming it mpLoadedModels, as the variable isn't a pointer itself. Gj 👍

Evayr commented on DirectX11Proj/d3dTexture.cpp:
```diff
@@ -9,4 +10,33 @@ d3dTexture::d3dTexture()
+}
+
+
+bool d3dTexture::Initialize(ID3D11Device* const aDevice, const WCHAR* aFilepath)
```
apDevice, apFilepath

Mercesa replied:
resolved

Evayr commented on DirectX11Proj/d3dTexture.cpp:
```diff
@@ -9,4 +10,33 @@ d3dTexture::d3dTexture()
+{
+	HRESULT result;
+
+	ID3D11ShaderResourceView* tShaderView = this->mpTexture.Get();
```
remove this->

Mercesa replied:
resolved

Evayr commented on DirectX11Proj/d3dTexture.cpp:
```diff
@@ -9,4 +10,33 @@ d3dTexture::d3dTexture()
+
+	ID3D11ShaderResourceView* tShaderView = this->mpTexture.Get();
+
+	result = D3DX11CreateShaderResourceViewFromFile(aDevice, aFilepath, NULL, NULL, this->mpTexture.GetAddressOf(), NULL);
```
remove this->

Mercesa replied:
resolved

Evayr commented on DirectX11Proj/d3dTexture.h:
```diff
@@ -1,10 +1,21 @@
 	d3dTexture();
 	~d3dTexture();
+
+	bool Initialize(ID3D11Device* const aDevice, const WCHAR* aFilepath);
```
apDevice, apFilepath

Mercesa replied:
resolved

Evayr commented on DirectX11Proj/d3dTexture.h:
```diff
@@ -1,10 +1,21 @@
+	bool Initialize(ID3D11Device* const aDevice, const WCHAR* aFilepath);
+	void Shutdown();
+
+	ID3D11ShaderResourceView* const GetTexture();
```
ID3D11ShaderResourceView* const GetTexture() **const**;

Evayr commented on DirectX11Proj/graphicsclass.cpp:
```diff
@@ -5,13 +5,13 @@
 {
 	mpDirect3D = 0;
-	mpModel = 0;
 	m_ColorShader = 0;
```
mColorShader

Mercesa replied:
mpColorShader now, resolved

Evayr commented on DirectX11Proj/graphicsclass.cpp:
```diff
@@ -62,32 +62,31 @@ bool GraphicsClass::Initialize(int screenWidth, int screenHeight, HWND hwnd)
-	if (!mpModel)
+
+	// Create the color shader object.
+	mTextureShader = std::make_unique<textureshaderclass>();
```
class without capital letters?

Evayr replied:
mpTextureShader

Mercesa replied:
legacy code, that's why no capitals. Will slowly convert in future

Evayr commented on DirectX11Proj/graphicsclass.cpp:
```diff
@@ -136,14 +135,19 @@ bool GraphicsClass::Render(IScene *const aScene)
-	{
-		return false;
+
+		if (aScene->mObjects[i]->mpModel->mHastexture)
```
apScene

Evayr replied:
Is an object a class? Call GetModel() and HasTexture() (or GetTextureCount()) instead. Also check for nullptrs

Mercesa replied:
Will be refactored, get model is a good one yes.

Evayr commented on DirectX11Proj/graphicsclass.cpp:
```diff
@@ -62,32 +62,31 @@ bool GraphicsClass::Initialize(int screenWidth, int screenHeight, HWND hwnd)
+	mTextureShader->Shutdown();
 	mpDirect3D->Shutdown();
 	
 	return;
```
Is this intentional?

Mercesa replied:
wasnt, resolved

Evayr commented on DirectX11Proj/modelclass.cpp:
```diff
@@ -1,8 +1,11 @@
-ModelClass::ModelClass()
+
+
+ModelClass::ModelClass() : mpTexture(nullptr)
```
new line if you want the fast declarations on a new line

Evayr commented on DirectX11Proj/modelclass.cpp:
```diff
@@ -1,8 +1,11 @@
 
 #include "d3dVertexBuffer.h"
+#include "d3dclass.h"
 #include "ModelData.h"
```
I'm a bit puzzled here. ModelData is using capital letters but d3dclass isn't?

Mercesa replied:
I use camel case for everything, d3dclass is partially legacy. renaming it is possible

Evayr commented on DirectX11Proj/modelclass.cpp:
```diff
@@ -21,26 +24,11 @@ ModelClass::~ModelClass()
-	return true;
-}
 
 bool ModelClass::Initialize(ID3D11Device* aDevice, const MeshData& aMesh)
```
apDevice, acMesh (c for const)

Evayr commented on DirectX11Proj/modelclass.h:
```diff
@@ -8,41 +8,34 @@ using namespace DirectX;
 
-	// Initialize buffers using previously loaded data
+	d3dTexture* mpTexture;
+	bool mHastexture = false;
```
make this private and make a get:
const bool HasTexture() const;

or
const uint8_t GetTextureCount() const;

Evayr commented on DirectX11Proj/modelclass.h:
```diff
@@ -8,41 +8,34 @@ using namespace DirectX;
+	bool mHastexture = false;
 
+private:
+	bool InitializeBuffers(ID3D11Device*, const MeshData&);
```
variable names?

Evayr commented on DirectX11Proj/modelclass.h:
```diff
@@ -8,41 +8,34 @@ using namespace DirectX;
+	bool InitializeBuffers(ID3D11Device*, const MeshData&);
 
 	void ShutdownBuffers();
 	void RenderBuffers(ID3D11DeviceContext*);
```
Variable names?

Mercesa replied:
fixed

Evayr commented on DirectX11Proj/systemclass.cpp:
```diff
@@ -100,6 +100,8 @@ bool SystemClass::Initialize()
 void SystemClass::Shutdown()
 {
 	// Release the graphics object.
+	ResourceManager::getInstance().Shutdown();
```
**G**etInstance()

Mercesa replied:
Fixed

Evayr commented on DirectX11Proj/systemclass.cpp:
```diff
@@ -100,6 +100,8 @@ bool SystemClass::Initialize()
 	// Release the graphics object.
+	ResourceManager::getInstance().Shutdown();
+
 	m_Graphics->Shutdown();
```
mpGraphics

Evayr commented on DirectX11Proj/systemclass.cpp:
```diff
@@ -100,6 +100,8 @@ bool SystemClass::Initialize()
+	ResourceManager::getInstance().Shutdown();
+
 	m_Graphics->Shutdown();
 	mApplication->Destroy();
```
mpApplication

Evayr commented on DirectX11Proj/texture.vs:
```diff
@@ -0,0 +1,47 @@
+{
+    float4 position : POSITION;
+    float2 uv : TEXCOORD0;
+	float3 normal : NORMAL;
```
indentation

Mercesa replied:
starting to believe there might be some parcing issues with github, not seeing this in vidusal studio

Evayr replied:
May be spaces instead of tabs

Evayr commented on DirectX11Proj/texture.vs:
```diff
@@ -0,0 +1,47 @@
+    output.position = mul(output.position, viewMatrix);
+    output.position = mul(output.position, projectionMatrix);
+        
+	output.uv = input.uv;
```
indentation

Evayr commented on DirectX11Proj/texture.vs:
```diff
@@ -0,0 +1,47 @@
+    PixelInputType output;
+    
+
+	// Change the position vector to be 4 units for proper matrix calculations.
```
are comments supposed to have one extra tab?

Mercesa replied:
no

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+
+textureshaderclass::textureshaderclass()
+{
+	m_vertexShader = 0;
```
mVertexShader, mPixelShader, etc

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+#include <iostream>
+#include "easylogging++.h"
+
+textureshaderclass::textureshaderclass()
```
No capital letters?

Mercesa replied:
legacy code

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+	m_layout = 0;
+	m_matrixBuffer = 0;
+
+	mpSampleState = 0;
```
Also, fast declare this

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+{
+}
+
+bool textureshaderclass::Initialize(ID3D11Device* device, HWND hwnd)
```
apDevice, aHWND

Evayr replied:
Maybe send HWND as const ref

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+	{
+		return false;
+	}
+
```
two enters here

Mercesa replied:
fixed

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+	if (mpSampleState)
+	{
+		mpSampleState->Release();
+		mpSampleState = 0;
```
= nullptr? I don't know if this is possible.

Mercesa replied:
me neither

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+}
+
+
+bool textureshaderclass::Render(ID3D11DeviceContext* deviceContext, int indexCount, XMMATRIX& worldMatrix, XMMATRIX& viewMatrix,
```
apDeviceContext, aIndexCount, aWorldMatric, aViewMatrix, ...

I know it's boring but it's not clean right now :(

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+
+
+bool textureshaderclass::Render(ID3D11DeviceContext* deviceContext, int indexCount, XMMATRIX& worldMatrix, XMMATRIX& viewMatrix,
+	XMMATRIX& projectionMatrix, ID3D11ShaderResourceView* texture)
```
apTexture

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+}
+
+
+bool textureshaderclass::InitializeShader(ID3D11Device* device, HWND hwnd, WCHAR* vsFilename)
```
apDevice, (const ref?) aHWND, apVSFilename

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+	HRESULT result;
+	ID3D10Blob* errorMessage;
+	ID3D10Blob* vertexShaderBuffer;
+	ID3D10Blob* pixelShaderBuffer;
```
you know what to do here

Mercesa replied:
si señor

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+	ID3D10Blob* errorMessage;
+	ID3D10Blob* vertexShaderBuffer;
+	ID3D10Blob* pixelShaderBuffer;
+	D3D11_INPUT_ELEMENT_DESC polygonLayout[5];
```
replace 5 with an enum

kPolygonLayoutCount or smth

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+	ID3D10Blob* vertexShaderBuffer;
+	ID3D10Blob* pixelShaderBuffer;
+	D3D11_INPUT_ELEMENT_DESC polygonLayout[5];
+	unsigned int numElements;
```
uint32_t

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+	D3D11_INPUT_ELEMENT_DESC polygonLayout[5];
+	unsigned int numElements;
+	D3D11_BUFFER_DESC matrixBufferDesc;
+	D3D11_SAMPLER_DESC samplerDesc;
```
naming

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+
+	// Create the vertex input layout description.
+	// This setup needs to match the VertexType structure in the ModelClass and in the shader.
+	polygonLayout[0].SemanticName = "POSITION";
```
polygonLayout[0].SemanticName = "POSITION";
polygonLayout[1].SemanticName = "TEXCOORD";
polygonLayout[2].SemanticName = "NORMAL";
polygonLayout[3].SemanticName = "BITANGENT";
polygonLayout[4].SemanticName = "TANGENT";

for (uint8_t i = 0; i < kPolygonLayoutCount; ++i
{
   polygonLayout[i].SemanticIndex = 0;
   polygonLayout[i].Format = DXGI_FORMAT_R32G32B32_FLOAT;
   polygonLayout[i].InputSlot = 0;
   polygonLayout[i].AlignedByteOffset = 0;
   polygonLayout[i].InputSlotClass = D3D11_INPUT_PER_VERTEX_DATA;
   polygonLayout[i].InstanceDataStepRate = 0;
}

Evayr replied:
they all seem to have the same values

Mercesa replied:
fair enough, good catch

Mercesa replied:
they're not all similar,sadly. The texcoord is a bit different from the rest. I'd prefer this way since otherwise itll not be clear to me what each attribute means

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+	polygonLayout[4].InstanceDataStepRate = 0;
+
+	// Get a count of the elements in the layout.
+	numElements = sizeof(polygonLayout) / sizeof(polygonLayout[0]);
```
kPolygonLayoutCount ?

Mercesa replied:
polygonlayout is an array

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+{
+	HRESULT result;
+	D3D11_MAPPED_SUBRESOURCE mappedResource;
+	MatrixBufferType* dataPtr;
```
pMatrixBuffer

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -0,0 +1,318 @@
+	HRESULT result;
+	D3D11_MAPPED_SUBRESOURCE mappedResource;
+	MatrixBufferType* dataPtr;
+	unsigned int bufferNumber;
```
uint32_t

Mercesa replied:
fixed

Evayr commented on DirectX11Proj/textureshaderclass.h:
```diff
@@ -0,0 +1,45 @@
+#pragma once
```
ifndef

Mercesa replied:
pragma once is fine too

Evayr commented on DirectX11Proj/textureshaderclass.h:
```diff
@@ -0,0 +1,45 @@
+using namespace DirectX;
+using namespace std;
+
+class textureshaderclass
```
capitals

Mercesa replied:
fixed

Evayr commented on DirectX11Proj/textureshaderclass.h:
```diff
@@ -0,0 +1,45 @@
+	ID3D11VertexShader* m_vertexShader;
+	ID3D11PixelShader* m_pixelShader;
+	ID3D11InputLayout* m_layout;
+	ID3D11Buffer* m_matrixBuffer;
```
mpMatrixBuffer, mpVertexShader, ...

Evayr commented on DirectX11Proj/textureshaderclass.h:
```diff
@@ -0,0 +1,45 @@
+	void ShutdownShader();
+
+	bool SetShaderParameters(ID3D11DeviceContext*, XMMATRIX&, XMMATRIX&, XMMATRIX&, ID3D11ShaderResourceView*);
+	void RenderShader(ID3D11DeviceContext*, int);
```
int32_t

Evayr replied:
argument names

Evayr commented on DirectX11Proj/textureshaderclass.h:
```diff
@@ -0,0 +1,45 @@
+	bool InitializeShader(ID3D11Device*, HWND, WCHAR*);
+	void ShutdownShader();
+
+	bool SetShaderParameters(ID3D11DeviceContext*, XMMATRIX&, XMMATRIX&, XMMATRIX&, ID3D11ShaderResourceView*);
```
argument names

Evayr commented on DirectX11Proj/textureshaderclass.h:
```diff
@@ -0,0 +1,45 @@
+	bool Render(ID3D11DeviceContext*, int, XMMATRIX&, XMMATRIX&, XMMATRIX&, ID3D11ShaderResourceView*);
+
+private:
+	bool InitializeShader(ID3D11Device*, HWND, WCHAR*);
```
argument names

Evayr commented on DirectX11Proj/textureshaderclass.h:
```diff
@@ -0,0 +1,45 @@
+
+	bool Initialize(ID3D11Device*, HWND);
+	void Shutdown();
+	bool Render(ID3D11DeviceContext*, int, XMMATRIX&, XMMATRIX&, XMMATRIX&, ID3D11ShaderResourceView*);
```
argument names, int32_t

Evayr commented on DirectX11Proj/textureshaderclass.h:
```diff
@@ -0,0 +1,45 @@
+	textureshaderclass();
+	~textureshaderclass();
+
+	bool Initialize(ID3D11Device*, HWND);
```
argument names

Mercesa replied:
fixed

Mercesa committed:
Textures and material system

Added support for multiple textures loading from within the model
loader.

Added general support for loading and using textures

Mercesa committed:
Added lighting

Normal mapping and multiple light support

Mercesa committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into niels

Mercesa committed:
Changed input system from xinput to windows messaging system

Mercesa committed:
Merge pull request #24 from Mercesa/InputRevision

Added new input

Evayr commented on DirectX11Proj/ConstantBuffers.h:
```diff
@@ -0,0 +1,2 @@
+#pragma once
```
ifndef

Mercesa replied:
pragma once is fine

Evayr commented on DirectX11Proj/Debug/DirectX11Proj.log:
```diff
@@ -1,4 +1,4 @@
-﻿  d3dTexture.cpp
```
remove this file & ignore it

Evayr commented on DirectX11Proj/Lighting.hlsl:
```diff
@@ -0,0 +1,123 @@
+	int padLB02;
+	int padLB03;
+
+
```
two enters - remove one

Mercesa replied:
fixed

Evayr commented on DirectX11Proj/Lighting.hlsl:
```diff
@@ -0,0 +1,123 @@
+	int padLB03;
+
+
+	Light arr[16];
```
kMaxLights

Mercesa replied:
Lighting.hlsl(36,12): error X3058: array dimensions must be literal scalar expressions

Evayr commented on DirectX11Proj/Lighting.hlsl:
```diff
@@ -0,0 +1,123 @@
+};
+// assume light at position 0,5,0 for now
+
+
```
two enters

Mercesa replied:
code is removed already, was deprecated

Evayr commented on DirectX11Proj/Lighting.hlsl:
```diff
@@ -0,0 +1,123 @@
+// assume light at position 0,5,0 for now
+
+
+float3 NormalSampleToWorldSpace(float3 normalMapSample, float3 unitNormalW, float3 tangentW)
```
aNormal, aUnit, etc.
Also const ref

Evayr commented on DirectX11Proj/Lighting.hlsl:
```diff
@@ -0,0 +1,123 @@
+	float3 reflV = reflect(-L, N);
+
+	// Calculate the specular intensity
+	float spec = pow(max(dot(V, reflV), 0.0f), 32);
```
32 -> kDefaultSomething

Evayr commented on DirectX11Proj/ModelLoader.cpp:
```diff
@@ -21,11 +21,13 @@ ModelLoader::~ModelLoader()
 
 std::string mDirectory;
```
global?

Evayr commented on DirectX11Proj/ModelLoader.cpp:
```diff
@@ -21,11 +21,13 @@ ModelLoader::~ModelLoader()
+
 void ModelLoader::ClearProcessedMeshes()
 {
 	this->mMeshesToBeProcessed.clear();
```
remove this->

Evayr commented on DirectX11Proj/ModelLoader.cpp:
```diff
@@ -21,11 +21,13 @@ ModelLoader::~ModelLoader()
 }
 
+
 void ProcessVertices(aiMesh* const a_Mesh, std::vector<VertexData>& a_Vertices)
```
apMesh

Evayr commented on DirectX11Proj/ModelLoader.cpp:
```diff
@@ -98,18 +100,19 @@ const std::vector<MeshData>& ModelLoader::GetMeshesToBeProcessed()
+	return true;
+}
 
+std::string GetTextureLocation(aiMaterial* const a_Mat, aiTextureType a_Type)
```
apMat

Evayr commented on DirectX11Proj/ModelLoader.cpp:
```diff
@@ -121,23 +124,38 @@ std::string GetTextureLocation(aiMaterial* const a_Mat, aiTextureType a_Type)
+
 // Goes through the material
-TextureData ProcessMaterial(aiMesh* a_Mesh, const aiScene* a_Scene)
+void ProcessMaterial(aiMesh* a_Mesh, const aiScene* a_Scene, MeshData& aMeshdata)
```
apMesh, apScene

Evayr commented on DirectX11Proj/ModelLoader.cpp:
```diff
@@ -121,23 +124,38 @@ std::string GetTextureLocation(aiMaterial* const a_Mat, aiTextureType a_Type)
 	if (a_Mesh->mMaterialIndex > 0)
 	{
-		aiMaterial* material = a_Scene->mMaterials[a_Mesh->mMaterialIndex];
+		aiMaterial* const material = a_Scene->mMaterials[a_Mesh->mMaterialIndex];
```
p prefix for pointer

Evayr commented on DirectX11Proj/PlayerSceneExample.cpp:
```diff
@@ -4,7 +4,8 @@
 #include "ObjectExample.h"
-
+#include <windows.h>
+#include <fcntl.h>
```
add enter

Evayr commented on DirectX11Proj/ResourceManager.cpp:
```diff
@@ -7,7 +7,7 @@
 #include "d3dTexture.h"
-
+#include "d3dMaterial.h"
 #include "easylogging++.h"
```
enter

Evayr commented on DirectX11Proj/ResourceManager.cpp:
```diff
@@ -16,52 +16,62 @@ ResourceManager::ResourceManager()
+	return std::move(tpTextureClass);
+}
+
+std::unique_ptr<d3dMaterial> ResourceManager::LoadTexturesFromMaterial(const MeshData& aMeshData)
```
acMeshData

Evayr commented on DirectX11Proj/ResourceManager.cpp:
```diff
@@ -16,52 +16,62 @@ ResourceManager::ResourceManager()
+}
+
 // TODO, create a function for model loader which just asks for the next model, this removes it from the model list
 std::vector<ModelClass*> ResourceManager::LoadModels(std::string aFilePath)
```
const ref str

Evayr commented on DirectX11Proj/d3dConstantBuffer.cpp:
```diff
@@ -0,0 +1,54 @@
+	aDevice->CreateBuffer(&CD3D11_BUFFER_DESC(aSize	, D3D11_BIND_CONSTANT_BUFFER, D3D11_USAGE_DYNAMIC, D3D11_CPU_ACCESS_WRITE), &d, &mpBuffer);
+}
+
+void d3dConstantBuffer::UpdateBuffer(void* data, ID3D11DeviceContext* const aContext)
```
apData apContext

Evayr commented on DirectX11Proj/d3dConstantBuffer.h:
```diff
@@ -0,0 +1,21 @@
+class d3dConstantBuffer
+{
+public:
+	d3dConstantBuffer(int aSize, void* aData, ID3D11Device* const aDevice);
```
apData, apDevice

Evayr commented on DirectX11Proj/d3dConstantBuffer.h:
```diff
@@ -0,0 +1,21 @@
+	d3dConstantBuffer(int aSize, void* aData, ID3D11Device* const aDevice);
+	~d3dConstantBuffer();
+	
+	void UpdateBuffer(void* data, ID3D11DeviceContext* const aContext);
```
apData, apContext

Evayr commented on DirectX11Proj/d3dTexture.cpp:
```diff
@@ -3,30 +3,31 @@
 #include <wrl.h>
 
-d3dTexture::d3dTexture()
+d3dTexture::d3dTexture() : exists(false)
```
new line, m_exists

Mercesa committed:
Did some cleaning

Renamed some files
Changed around camera controls
Fixed release mode

Evayr commented on DirectX11Proj/inputclass.cpp:
```diff
@@ -5,9 +5,9 @@
-	this->mpMouse = 0;
+	//this->mpDirectInput = 0;
+	//this->mpKeyboard = 0;
+	//this->mpMouse = 0;
```
delete this

Evayr commented on DirectX11Proj/inputclass.cpp:
```diff
@@ -34,68 +34,9 @@ bool InputClass::Initialize(HINSTANCE hinstance, HWND hwnd, int screenWidth, int
 
-	result = DirectInput8Create(hinstance, DIRECTINPUT_VERSION, IID_IDirectInput8, (void**)&mpDirectInput, NULL);
-	if (FAILED(result))
+	for (int i = 0; i < 256; ++i)
```
kSomething

Evayr commented on DirectX11Proj/inputclass.cpp:
```diff
@@ -104,122 +45,36 @@ bool InputClass::Initialize(HINSTANCE hinstance, HWND hwnd, int screenWidth, int
+	//if (mMouseY < 0) { mMouseY = 0; }
+	//
+	//if (mMouseX > mScreenWidth) { mMouseX = mScreenWidth; }
+	//if (mMouseY > mScreenHeight) { mMouseY = mScreenHeight; }
```
remove this (unless WIP)

Evayr commented on DirectX11Proj/inputclass.h:
```diff
@@ -22,26 +22,49 @@ class InputClass
 	int mMouseX, mMouseY;
 	int mMouseRelX, mMouseRelY;
 
+	bool mKeys[256];
```
kSomething

Evayr replied:
maybe have key hit as well (instead of up/down)

Evayr commented on DirectX11Proj/modelclass.h:
```diff
@@ -11,29 +11,30 @@ class d3dVertexBuffer;
 	~ModelClass();
 
-	bool Initialize(ID3D11Device*, const MeshData&);
+	bool Initialize(ID3D11Device* const apDevice, const MeshData& aMeshData);
```
acMeshData

Evayr commented on DirectX11Proj/modelclass.h:
```diff
@@ -11,29 +11,30 @@ class d3dVertexBuffer;
 
 private:
-	bool InitializeBuffers(ID3D11Device*, const MeshData&);
+	bool InitializeBuffers(ID3D11Device* const apDevice, const MeshData& aMeshData);
```
acMeshData

Evayr commented on DirectX11Proj/systemclass.cpp:
```diff
@@ -184,24 +178,55 @@ bool SystemClass::Frame()
-	mApplication->Tick();
-	mApplication->SceneTick(m_Input.get());
+	mpApplication->Tick();
+	mpApplication->SceneTick(m_Input.get());
```
mpInput

Evayr commented on DirectX11Proj/systemclass.cpp:
```diff
@@ -184,24 +178,55 @@ bool SystemClass::Frame()
 }
 
 
 LRESULT CALLBACK SystemClass::MessageHandler(HWND hwnd, UINT umsg, WPARAM wparam, LPARAM lparam)
```
arg names

Evayr commented on DirectX11Proj/texture.vs:
```diff
@@ -33,15 +39,33 @@ PixelInputType TextureVertexShader(VertexInputType input)
+	output.fragPos = output.position;
+	
+	output.norm = mul(float4(input.normal, 1.0), worldMatrix);
+	output.tang = mul(float4(input.tangent, 1.0), worldMatrix);
```
indentation

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -55,45 +54,55 @@ void textureshaderclass::ShutdownShader()
 
 bool textureshaderclass::Render(ID3D11DeviceContext* deviceContext, int indexCount, XMMATRIX& worldMatrix, XMMATRIX& viewMatrix,
-	XMMATRIX& projectionMatrix, ID3D11ShaderResourceView* texture)
+	XMMATRIX& projectionMatrix, d3dMaterial* const aMaterial, std::vector<unique_ptr<Light>>& aLight, XMFLOAT3 aCamPos)
```
acMat

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -113,7 +122,7 @@ bool textureshaderclass::InitializeShader(ID3D11Device* device, HWND hwnd, WCHAR
 	ID3D10Blob* vertexShaderBuffer;
 	ID3D10Blob* pixelShaderBuffer;
```
pPixel, pVertex

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -256,21 +299,86 @@ bool textureshaderclass::InitializeShader(ID3D11Device* device, HWND hwnd, WCHAR
 	return true;
 }
 
+bool textureshaderclass::SetLightConstantBufferData(ID3D11DeviceContext* const aDeviceContext, std::vector<unique_ptr<Light>>& aLights)
```
apDevice

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -256,21 +299,86 @@ bool textureshaderclass::InitializeShader(ID3D11Device* device, HWND hwnd, WCHAR
+{
+	HRESULT result;
+	D3D11_MAPPED_SUBRESOURCE mappedResource;
+	LightBufferType* dataPtr;
```
pData

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -256,21 +299,86 @@ bool textureshaderclass::InitializeShader(ID3D11Device* device, HWND hwnd, WCHAR
+	aDeviceContext->PSSetConstantBuffers(bufferNumber, 1, &mpLightConstantBuffer);
+}
+
+bool textureshaderclass::SetMaterialConstantBufferData(ID3D11DeviceContext* const aDeviceContext, d3dMaterial* const aMaterial)
```
apDevice, apMaterial

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -256,21 +299,86 @@ bool textureshaderclass::InitializeShader(ID3D11Device* device, HWND hwnd, WCHAR
+{
+	HRESULT result;
+	D3D11_MAPPED_SUBRESOURCE mappedResource;
+	MaterialBufferType* dataPtr;
```
pData

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -283,36 +391,52 @@ bool textureshaderclass::SetShaderParameters(ID3D11DeviceContext* deviceContext,
+	deviceContext->PSSetShaderResources(1, 1, &aView2);
+
+
+	ID3D11ShaderResourceView* aView3 = aMaterial->mpNormal->GetTexture();
```
apView3

Evayr commented on DirectX11Proj/textureshaderclass.cpp:
```diff
@@ -283,36 +391,52 @@ bool textureshaderclass::SetShaderParameters(ID3D11DeviceContext* deviceContext,
 	return true;
 }
 
 void textureshaderclass::RenderShader(ID3D11DeviceContext* deviceContext, int indexCount)
```
apDevice, aIndex

Evayr commented on DirectX11Proj/textureshaderclass.h:
```diff
@@ -18,6 +23,25 @@ class textureshaderclass
+		int amountOfLights;
+		int padLB01;
+		int padLB02;
+		int padLB03;
```
int32_t for all

Mercesa committed:
Constant buffer introduced

Introduced constant buffer in texture shader class

Mercesa committed:
Revamped shader system

A shader management system has been set in place

Mercesa committed:
Added too many thing's can't keep track

Mercesa committed:
Added some new classes, courtesy of Rastertek d3d11 tutorials

Mercesa committed:
Added directxTK and shaders for shadows

Mercesa committed:
Refactored some code, added a new global managing class

Mercesa committed:
Worked on systems
